### PR TITLE
Use intrinsics for base integer operations

### PR DESCRIPTION
**Pull request by IGI-111**
## Description
This change introduces `__mod`, `__shl` and `__shr` intrinsics for use with integer operations to remove the asm block based implementations in core and enable us to support const evaluation of integer expressions in most cases, notably when specifying values like `1 << 32`.

## Checklist

- [x] I have linked to any relevant issues.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have updated the documentation where relevant (API docs, the reference, and the Sway book).
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] I have added (or requested a maintainer to add) the necessary `Breaking*` or `New Feature` labels where relevant.
- [x] I have done my best to ensure that my PR adheres to [the Fuel Labs Code Review Standards](https://github.com/FuelLabs/rfcs/blob/master/text/code-standards/external-contributors.md).
- [x] I have requested a review from the relevant team or maintainers.

### CODE REVIEW

**Review by vaivaswatha — COMMENTED** [review]
The following need to be done:
1. Support for parsing the new IR instructions in `sway-ir/src/parser.rs`. Consequently, those must be tested via IR parser tests in `tests/ir_to_asm`.
2. Unit test for `const_eval` additions in `should_pass/language/const_inits/src/main.sw`.
3. Unit tests for the intrinsics in  `should_pass/language/binop_intrinsics/src/main.sw`.

References:
* https://github.com/FuelLabs/sway/pull/2771/files
* https://github.com/FuelLabs/sway/pull/4260/files (This one is missing the IR parser tests in `tests/ir_to_asm` as well. If you could add those, that would be even better. that was my bad missing them).

**Comment by vaivaswatha** [comment]
> > IR parser tests in `tests/ir_to_asm`.
> 
> For the record, those have apparently been moved to `test/src/ir_generation`

Looks like they were removed, not moved.

**Comment by IGI-111** [comment]
There's still one test in there for `__eq`, I'll just expand that to test all the intrinsics.

**Comment by vaivaswatha** [comment]
> > > IR parser tests in `tests/ir_to_asm`.
> > 
> > 
> > For the record, those have apparently been moved to `test/src/ir_generation`
> 
> Looks like they were removed, not moved.

Which makes me wonder what's the best place to have IR parsing tests. Probably `sway-ir/tests`. We could have an Issue to add IR tests for all intrinsics.

**Review by vaivaswatha — DISMISSED** [review]

File: test/src/ir_generation/binops.sw [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+    let _ = __mul(1, 2);
+    let _ = __div(1, 2);
+    let _ = __and(1, 2);
+    let _ = __or(1, 2);
+    let _ = __xor(1, 2);
+    let _ = __mod(1, 2);
+    let _ = __rsh(1, 2);
+    let _ = __lsh(1, 2);
+}
+// check: fjkldjslfjkdl
```
[the remark this answers is not in the document]
Comment (IGI-111) [reply test/src/ir_generation/binops.sw]: I was puzzled as to how this was still in with the test passing but turns out when I cherry-picked my test branch I moved it in the wrong folder.
Should be fixed now.

**Review by vaivaswatha — APPROVED** [review]

**Pull request merged**